guvnor-ala/guvnor-ala-spi/src/main/java/org/guvnor/ala/pipeline/execution/PipelineExecutor.java

```diff
@@ -37,75 +37,77 @@
                              final Consumer<T> callback,
                              final PipelineEventListener... eventListeners ) {
-        final PipelineContext context = new PipelineContext( pipeline );
+        final PipelineContext context = new PipelineContext(pipeline);
         context.start( input );
         context.pushCallback( callback );
-        propagateEvent( new BeforePipelineExecutionEvent( pipeline ), eventListeners );
+        propagateEvent(new BeforePipelineExecutionEvent(context.getExecutionId(), pipeline ), eventListeners );
         continuePipeline( context, eventListeners );
-        propagateEvent( new AfterPipelineExecutionEvent( pipeline ), eventListeners );
+        propagateEvent(new AfterPipelineExecutionEvent(context.getExecutionId(), pipeline ), eventListeners );
     }
 
     private void continuePipeline( final PipelineContext context,
                                    final PipelineEventListener... eventListeners ) {
         while ( !context.isFinished() ) {
-            final Stage<Object, ?> stage = getCurrentStage( context );
+            final Stage<Object, ?> stage = getCurrentStage(context );
             final Object newInput = pollOutput( context );
 
             try {          
-                propagateEvent( new BeforeStageExecutionEvent( context.getPipeline(), stage ), eventListeners );
+                propagateEvent(new BeforeStageExecutionEvent(context.getExecutionId(), context.getPipeline(), stage ), eventListeners );
                 stage.execute( newInput, output -> {
                     
-                    final ConfigExecutor executor = resolve( output.getClass() );
-                    if ( output instanceof ContextAware ) {
-                        ( (ContextAware) output ).setContext( Collections.unmodifiableMap( context.getValues() ) );
+                    final ConfigExecutor executor = resolve(output.getClass() );
+                    if ( output instanceof ContextAware) {
+                        ( (ContextAware) output ).setContext(Collections.unmodifiableMap(context.getValues() ) );
                     }
                     final Object newOutput = interpolate( context.getValues(), output );
                     if(executor == null){
                         throw new RuntimeException("Fail to resolve ConfigExecutor for: " + output.getClass());
                     }
                     context.getValues().put( executor.inputId(), newOutput );
-                    if ( executor instanceof BiFunctionConfigExecutor ) {
-                        final Optional result = (Optional) ( (BiFunctionConfigExecutor) executor ).apply( newInput, newOutput );
+                    if ( executor instanceof BiFunctionConfigExecutor) {
+                        final Optional result = (Optional) ( (BiFunctionConfigExecutor) executor ).apply(newInput, newOutput );
                         context.pushOutput( executor.outputId(), result.get() );
```

Comment:
This looks like using optional.get without `isPresent()` in which case an exception will be thrown.